### PR TITLE
Switched to using the `tiged` module instead of the `degit` module.

### LINKED ISSUE #28: Cannot fetch gitlab project
Attempting to use `tmplr` on my gitlab project doesn't work, it simply says `Working` and spins forever. Using `local` to the checked out project works perfectly, as does `tmplr preview`.

I believe this is related to issues `degit` has with gitlab and projects in subgroups; I had to switch to using `tiged` instead (mainly because of this [issue](https://github.com/tiged/tiged/pull/24)). (I found `tmplr` from a `tiged` issue, actually, and it's exactly what I want, love it so far!)

My project is here: https://gitlab.com/skewed-aspect/seed-projects/vue3-bootstrap (currently, the tmplr implementation is in the `template` branch, but `tmplr` failed on the base project).

I've tried the following commands, and they all hang:

```bash
$ npx tmplr https://gitlab.com/skewed-aspect/seed-projects/vue3-bootstrap#template
$ npx tmplr https://gitlab.com/skewed-aspect/seed-projects/vue3-bootstrap
```

Screenshot:
image

### PR DESCRIPTION
**Pull request by Morgul**
This change switches to using `tiged` in place of `degit`, to fix issues with cloning from gitlab.

This should fix https://github.com/loreanvictor/tmplr/issues/28.

Added types from `@types/degit`, ported to match `tiged`'s api. (Note: There's an open [PR](https://github.com/tiged/tiged/pull/89) porting `tiged` to TypeScript, so these types won't be required forever.)

### CODE REVIEW

**Comment by loreanvictor** [comment]
@Morgul I think some tests need to be updated. additionally, [degit's issue with subgroups are supposed to be fixed now (?)](https://github.com/tiged/tiged/pull/24). should we still replace degit with tiged?

**Comment by Morgul** [comment]
So, the issues with the tests are cause because I forked the project, so for this test, it's getting a different value:

image

I didn't think updating it made since, as it would break the main repo once it was merged. I'm happy to update the test, just not sure the best way to update. Maybe we just skip getting the repo owner as part of the test?

As for `degit`: I still can't use tmplr with subgroups on gitlab, so I think the change is still needed. It hasn't had any commits in 3 years and I think most development is happening over at `tiged` anyway.

**Comment by loreanvictor** [comment]
> So, the issues with the tests are cause because I forked the project, so for this test, it's getting a different value ...
> I'm happy to update the test, just not sure the best way to update. Maybe we just skip getting the repo owner as part of the test?
> 

Well then we wouldn't be testing that functionality 😅 Yeah this is my bad because I didn't think of forks when writing this particular test. I suppose the cleanest solution would be to create a temporary folder (this is already done in the next test), clone a pre-designated repository inside it, and then create a git provider within that repository and check the values accordingly, so the test is not dependent on the current repository.

If fixing this issue is a hassle though, I suppose it is ok to ignore the test right now and issue a fix for it later, since it wouldn't fail for the actual build of the library.

Also would it be possible to add a test highlighting the issue you had? e.g. using `node-fs` to clone a repository that couldn't have been cloned before?

**Comment by Morgul** [comment]
> If fixing this issue is a hassle though, I suppose it is ok to ignore the test right now and issue a fix for it later, since it wouldn't fail for the actual build of the library.

I'll give it a shot, I should be able to work that out. I was mainly trying to have as minimally invasive a set of changes as possible.

> Also would it be possible to add a test highlighting the issue you had? e.g. using `node-fs` to clone a repository that couldn't have been cloned before?

...Ok, this one's my bad. I should have added a test for that! Probably won't be today, but I'll circle back around and get this updated.

**Comment by Morgul** [comment]
> I'll give it a shot, I should be able to work that out. I was mainly trying to have as minimally invasive a set of changes as possible.

Pushed a change that makes a temp directory and does a checkout. I went with the simplest option, which is just using `child-process` to call `git`; I wasn't sure if you'd rather add something like `node-git` or how you'd want to handle it, but it seems to work well.

I'll work on writing the new test next, figured I'd get this change in first so you could look at it and give me feedback if you want changes.

**Comment by loreanvictor** [comment]
> > I'll give it a shot, I should be able to work that out. I was mainly trying to have as minimally invasive a set of changes as possible.
> 
> Pushed a change that makes a temp directory and does a checkout. I went with the simplest option, which is just using `child-process` to call `git`; I wasn't sure if you'd rather add something like `node-git` or how you'd want to handle it, but it seems to work well.
> 
> I'll work on writing the new test next, figured I'd get this change in first so you could look at it and give me feedback if you want changes.

the code looks good to me. I was wondering whether there the OS-dependent quirks of `childProcess()` would affect the test in different environments but it appears not (as all tests are passing correctly in all envs).

**Comment by Morgul** [comment]
@loreanvictor Sorry it's taken me a minute to finish this up.

I've got a test all worked up for the new functionality, but I've hit a snag I missed previously; it looks like you need to pass `tiged` a `subgroup: true` option to make it work correctly with subgroups. Doing that works well, but breaks anything not a subgroup.

Do you think it's worth adding support for setting tiged options through the `fetch` function? Or maybe just adding a single optional `subgroup` parameter to control the required tiged option?

**Comment by loreanvictor** [comment]
> @loreanvictor Sorry it's taken me a minute to finish this up.
> 
> I've got a test all worked up for the new functionality, but I've hit a snag I missed previously; it looks like you need to pass `tiged` a `subgroup: true` option to make it work correctly with subgroups. Doing that works well, but breaks anything not a subgroup.
> 
> Do you think it's worth adding support for setting tiged options through the `fetch` function? Or maybe just adding a single optional `subgroup` parameter to control the required tiged option?

what are other tiged options? I suspect we could add an option for subgroups, up-propagating it to `tmplr` CLI itself. it's not really neat, but it might be ok.

**Comment by Morgul** [comment]
For now, I exposed the subgroup option. The other options seem like renamed versions of the `degit` options, so it's probably not worth exposing all of them right now.

File: src/filesystem/node-fs.ts [hunk 1]
```diff
@@ -101,7 +101,7 @@ export class NodeFS implements FileSystem {
     return new NodeFS(this.scope, abs)
   }
 
-  async fetch(remote: string, dest: string) {
+  async fetch(remote: string, dest: string, subgroup = false) {
```
Comment (loreanvictor) [inline src/filesystem/node-fs.ts]: @Morgul could you use an optional `options` object? adding positional arguments is not really scalable after one or two. in this case, if we add another option in the future, then providing that option would require users to know and provide `subgroup` as well, increasing usage hassle.

**Review by loreanvictor — CHANGES_REQUESTED** [review]

File: src/filesystem/node-fs.ts (hunk 1 above)
Comment (Morgul) [reply src/filesystem/node-fs.ts]: Sure, that makes complete sense. Just pushed a version with an options object.

**Review by loreanvictor — APPROVED** [review]
Looks good. I'll merge this, test locally and publish a new version. we can subsequently upgrade dependencies in dependent packages. I think we'd additionally need parser options for "degit" command, alongside CLI options to use those parser options. I'll create the corresponding issues on relevant repos as well.

**Comment by loreanvictor** [comment]
as a reminder for future, and to track progress, putting this on production requires:

- [ ] adding the newly added options to the [`filesystem` type](https://github.com/loreanvictor/tmplr-core/blob/0ca3badc23658583da9cdedcfdb032256376b7a0/src/filesystem/types.ts#L15)
- [ ] adding `subgroup` option for [`degit` command](https://github.com/loreanvictor/tmplr-core/blob/main/src/command/degit.ts)
- [ ] publishing a new version of `tmplr-core`
- [ ] updating `tmplr-node` to use the new core, publish a new version of `tmplr-node`
- [ ] updating the [yaml parser](https://github.com/loreanvictor/tmplr-yaml-parser) to use the new core, and adding support for parsing the new subgroup option in the [corresponding parser rule in YAML parser](https://github.com/loreanvictor/tmplr-yaml-parser/blob/main/src/core/command/degit.ts).
- [ ] publishing the new parser.
- [ ] updating core version on all `tmplr-*` packages, publishing new versions when necessary.
- [ ] updating [`tmplr`](https://github.com/loreanvictor/tmplr) proper to use the new packages. adding support for the new subgroup option to [the corresponding recipe](https://github.com/loreanvictor/tmplr/blob/main/src/recipes/degit-and-run.ts), and to the [CLI args parser](https://github.com/loreanvictor/tmplr/blob/main/src/app/args.ts).
- [ ] publishing the new tmplr!

**Comment by Morgul** [comment]
Thanks for that checklist, it means (as I have time) I can work on the various parts of this to get it across the line!

**Pull request merged**